File: atcoder.jp/contests/abc270/tasks/abc270_d/main.py
from itertools import takewhile

N, K = map(int, input().split())
A = list(map(int, input().split()))

dp = [0] * (N + 1)

for n in range(1, N + 1):
    dp[n] = max(n - dp[n - a] for a in takewhile(lambda a: a <= n, A))

ans = dp[N]
print(ans)

File: atcoder.jp/contests/abc270/tasks/abc270_d/main.cpp
#include <bits/stdc++.h>
using namespace std;

template <typename T>
void print(T value) {
  cout << value << endl;
}

vector<int> range(int start, int stop) {
  vector<int> ret(stop - start);
  iota(ret.begin(), ret.end(), start);
  return ret;
}

vector<int> input_vector(int n) {
  vector<int> ret(n);
  for (auto &&i : ret) cin >> i;
  return ret;
}

int main() {
  int N, K;
  cin >> N >> K;
  vector<int> A = input_vector(K);

  vector<int> dp(N + 1);

  for (auto &&n : range(1, N + 1)) {
    for (auto &&a : A) {
      if (n < a) break;
      dp[n] = max(dp[n], n - dp[n - a]);
    }
  }

  int ans = dp[N];
  print(ans);
}
